### PR TITLE
Add multinic tests to kitchen.recipes.yml

### PR DESCRIPTION
**Pull request by awslitvit**
### Description of changes
* [Enable kitchen lifecycle hooks for recipes](https://github.com/aws/aws-parallelcluster-cookbook/commit/21e8ead3f44b9ac20082f859f47cd88ac8daabe8) 
* [Add multinic tests to kitchen.recipes.yml](https://github.com/aws/aws-parallelcluster-cookbook/commit/06ffa28d37b337a7f9b67264c9dca9c3ae8b69fe)

### Tests
* `network_interfaces` run manually on EC2

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb [hunk 1]
```diff
@@ -21,6 +21,27 @@
   end
 end
 
+control 'multiple_network_interfaces_can_ping_gateway' do
```
Comment (francesco-giordano) [inline test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:24]: Suggestion: can you add some comments to clarify what this test checks?

File: kitchen.recipes.yml [hunk 2]
```diff
@@ -15,6 +15,12 @@ verifier:
     - test/recipes
     - test/resources
 
+lifecycle:
+  <% %w(post_create pre_converge pre_destroy).each do |op| %>
```
Comment (francesco-giordano) [inline kitchen.recipes.yml:19]: Question: Which are the other possible operations? If they are only 3 would not be clearer to have 6 lines?

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb [hunk 3]
```diff
@@ -21,6 +21,27 @@
...
+  only_if { !os_properties.virtualized? }
+
+  describe bash("ip -o link | grep -v lo | wc -l") do
+    its('stdout.strip') { should cmp >= 2 }
+  end
+
+  gateway_ip = bash("ip r | grep default | head -n 1 | awk '{print $3}'").stdout.strip
+  ips = bash("ip r | grep src | awk '{ print $9}'").stdout.split(/\n+/)
```
Comment (francesco-giordano) [inline test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb]: Typo: space in bracket

Issue: with `ip route` we only get the IP from the routing table, if we misconfigure the routing table we could miss errors. I think would be better to use `ip address`

File: test/recipes/hooks/network_interfaces/post_create.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+echo "**** post_create"
+
+[[ "${KITCHEN_DRIVER}" = "ec2" ]] || return
+
+echo "** Creating ENI..."
+
+KITCHEN_ENI_ID=$(aws ec2 create-network-interface --region "${KITCHEN_AWS_REGION}" \
+    --description "Kitchen tests ENI for ${KITCHEN_SUITE_NAME} on ${KITCHEN_EC2_INSTANCE_ID}" \
```
Comment (francesco-giordano) [inline test/recipes/hooks/network_interfaces/post_create.sh:10]: Question: where is it defined `KITCHEN_SUITE_NAME `? Is it set in the environment by kitchen?

File: kitchen.recipes.yml [hunk 5]
```diff
@@ -15,6 +15,12 @@ verifier:
     - test/recipes
     - test/resources
 
+lifecycle:
```
Comment (francesco-giordano) [inline kitchen.recipes.yml:18]: Question (non-blocking): is the lifecycle executable in parallel?

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb [hunk 6]
```diff
@@ -42,7 +63,7 @@
         its('group') { should eq 'root' }
         its('content') { should match /^ip route del/ }
       end
-    else
+    elsif os_properties.amazon_family? || os_properties.centos?
```
Comment (francesco-giordano) [inline test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:67]: Question: is the change to make the branch explicit?

File: test/recipes/hooks/run.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    KITCHEN_EC2_INSTANCE_ID=$(ssh -o StrictHostKeyChecking=no -i "${KITCHEN_SSH_KEY_PATH}" \
+      "${KITCHEN_EC2_USER}@${KITCHEN_INSTANCE_HOSTNAME}" '
+      TOKEN=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600") \
+      && curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/instance-id
+    ')
+
+    echo "EC2 instance id: ${KITCHEN_EC2_INSTANCE_ID}"
+
+    export KITCHEN_EC2_INSTANCE_ID
```
Comment (francesco-giordano) [inline test/recipes/hooks/run.sh:29]: Question (non-blocking): Related to, is it executable in parallel?

File: test/recipes/hooks/run.sh [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+export KITCHEN_HOOK=$1
+export KITCHEN_ROOT_DIR=$2
+
+echo "*** Run ${KITCHEN_HOOK} for suite ${KITCHEN_SUITE_NAME} on host ${KITCHEN_INSTANCE_HOSTNAME}"
```
Comment (francesco-giordano) [inline test/recipes/hooks/run.sh:6]: Question: where is it defined `KITCHEN_INSTANCE_HOSTNAME` ? Is it set in the environment by kitchen?

File: kitchen.recipes.yml (hunk 2 above)
Comment (awslitvit) [reply kitchen.recipes.yml:19]: We can potentially add other operations in the future: these are the operations I've used in my tests so far.
Having this loop is less error prone.

File: test/recipes/hooks/network_interfaces/post_create.sh (hunk 4 above)
Comment (awslitvit) [reply test/recipes/hooks/network_interfaces/post_create.sh:10]: Yes, it is handled by kitchen: see [here](https://kitchen.ci/docs/reference/lifecycle-hooks/).

File: kitchen.recipes.yml (hunk 5 above)
Comment (awslitvit) [reply kitchen.recipes.yml:18]: Do you mean in parallel across suites? Yes, it is.

File: test/recipes/hooks/run.sh (hunk 7 above)
Comment (awslitvit) [reply test/recipes/hooks/run.sh:29]: Yes, I tested it in parallel for multiple OSes.

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb (hunk 6 above)
Comment (awslitvit) [reply test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:67]: This test doesn't work on RHEL8. I thought of removing it at all, then I opted for excluding just RHEL8.

File: test/recipes/hooks/run.sh (hunk 8 above)
Comment (awslitvit) [reply test/recipes/hooks/run.sh:6]: Yes, same as above.

**Review by francesco-giordano — APPROVED** [review]

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb (hunk 1 above)
Comment (awslitvit) [reply test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:24]: I believe the control name is clean to this end. If it is not, please tell me what you want to be better clarified.

File: test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb (hunk 3 above)
Comment (awslitvit) [reply test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb]: Done.

**Pull request merged**